### PR TITLE
Test driven development branch

### PR DESCRIPTION
**Pull request by andru47**
Add some tests for the Greeter and modify Greeter to remove symbols
Add code for mandatory and optional participants
Add tests for the new features

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,57 @@
...
+        busyTimesBoth.add(it.getWhen());
+      if (!eventHasGivenParticipants(it.getAttendees(), mandatoryParticipants)) // Do not include event if no mandatory participants attend
+        continue;
+      busyTimesMandatory.add(it.getWhen());
+    }
+
+    Collection<TimeRange> availableTimesForBoth = getAvailableTimes(busyTimesBoth, request.getDuration());
+    if (availableTimesForBoth.isEmpty())
+      return getAvailableTimes(busyTimesMandatory, request.getDuration());
+    else return availableTimesForBoth;
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:43]: Can you omit the else here?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,57 @@
...
+  }
+
+  private Collection<TimeRange> getAvailableTimes(Collection<TimeRange> busyTimes, long duration){
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: It's not clear from the name what time ranges are being returned, please document.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,57 @@
...
+    if (busyTimes.isEmpty() && duration <= TimeRange.WHOLE_DAY.duration())
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    int last = TimeRange.START_OF_DAY;
+    
+    for (TimeRange it : busyTimes){
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: Why is it called `it`?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,57 @@
...
+    if (availableTimesForBoth.isEmpty())
+      return getAvailableTimes(busyTimesMandatory, request.getDuration());
+    else return availableTimesForBoth;
+  }
+
+  private Collection<TimeRange> getAvailableTimes(Collection<TimeRange> busyTimes, long duration){
+    if (busyTimes.isEmpty() && duration <= TimeRange.WHOLE_DAY.duration())
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    int last = TimeRange.START_OF_DAY;
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: Is `last` the correct name for this variable? I'm not entirely sure if those fit everywhere, but alternatives are `current` or `start_candidate` .

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,57 @@
...
+    
+    for (TimeRange it : busyTimes){
+      if (it.start() >= last && it.start() - last >= duration)
+        availableTimes.add(TimeRange.fromStartEnd(last, it.start(), false));
+      if (last< it.end())
+        last = it.end();
+    }
+    if (TimeRange.END_OF_DAY -  last >= duration)
+      availableTimes.add(TimeRange.fromStartEnd(last, TimeRange.END_OF_DAY, true));
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: The definition of END_OF_DAY or of the range is confusing here. For time ranges, `end` is the first minute that is not included in the range -- why is END_OF_DAY included in the last interval then?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,57 @@
...
+
+    return availableTimes;
+  }
+
+  private boolean eventHasGivenParticipants(Collection<String> eventParticipants, Collection<String> givenParticipants){
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: It's not clear from the name if its "HasAllOf" or "HasAny"

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 7]
```diff
@@ -44,6 +44,12 @@ public MeetingRequest(Collection<String> attendees, long duration) {
     this.attendees.addAll(attendees);
   }
 
+  public MeetingRequest(Collection<String> attendees, Collection<String> optional_attendees, long duration){
+    this.duration = duration;
+    this.attendees.addAll(attendees);
+    this.optional_attendees.addAll(optional_attendees);
```
Comment (pmueller-gg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:50]: Nit: Why does the order of assignments and arguments differ?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (andru47) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: I just saw in their tests that true should be set for END_OF_DAY

Comment (pmueller-gg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Ok.

**Pull request merged**